// botbuilder-quickreplies/index.js
/*-----------------------------------------------------------------------------
BotBuilder middleware and utility classes for supporting facebook quick replies
when using a LUIS recognizer.
-----------------------------------------------------------------------------*/
var builder = require('botbuilder');
var locationPrompt = require('./location-prompt');

/*-----------------------------------------------------------------------------
Quick Replies Middleware
 - Checks to see if the incoming message was triggered by a facebook quick
   reply, and sets the message text to be the payload value to allow LUIS
   or dialogs to use the correct value.
-----------------------------------------------------------------------------*/
var QuickRepliesMiddleware = {
    botbuilder: function (session, next) {
        if (session.message.source == 'facebook') {
            if (session.message.sourceEvent && session.message.sourceEvent.message) {
                if (session.message.sourceEvent.message.quick_reply) {
                    session.message.text = session.message.sourceEvent.message.quick_reply.payload;
                }        
            }
        }

        next();
    }
}

/*-----------------------------------------------------------------------------
Utility for constructing a location quick reply.
-----------------------------------------------------------------------------*/
function QuickReplyLocation() {
    return {
        content_type:"location",
    }
}

/*-----------------------------------------------------------------------------
Utility for constructing a quick reply.
-----------------------------------------------------------------------------*/
function QuickReplyText(title, payloadString, imgUrl) {
    var quickReply = {
        content_type:"text",
        title: title,
        payload: payloadString
    }

    if (imgUrl) {
        quickReply.image_url = imgUrl;
    }

    return quickReply;
}

/*-----------------------------------------------------------------------------
Utility for adding specified replys to a bot builder message.
-----------------------------------------------------------------------------*/
function AddQuickReplies(session, message, quickReplys) {
    if (session.message.source == 'facebook') {
        message.sourceEvent({
            facebook: {
                quick_replies: quickReplys
            }
        })
    }

    return message;
}

exports.QuickRepliesMiddleware = QuickRepliesMiddleware;
exports.QuickReplyText = QuickReplyText;
exports.QuickReplyLocation = QuickReplyLocation;
exports.AddQuickReplies = AddQuickReplies;
exports.LocationPrompt = locationPrompt;
